### PR TITLE
Inference module updated to Pyhton 3.10.6.

### PR DESCRIPTION
**Pull request by SMoragon**
Main changes:

- Added input variable and return types to all the methods.

- Fixed some imports.

- Changed the way to access dictionaries.

The main aim of this is to fulfill the version update. 
Please, be aware that it's still under revision and there might be further changes in future versions.

### CODE REVIEW

File: mevolib/inference/__init__.py [hunk 1]
```diff
@@ -89,15 +106,15 @@ def get_keywords ( tool ) :
...
         keyword_dict[key] = ' '.join(value)
-    return ( keyword_dict )
+    return ( keyword_dict ) 
 
 
 
-def get_phylogeny ( binary, infile, infile_format, args = 'default',
-                    outfile = None, outfile_format = 'newick',
-                    bootstraps = 0 ) :
+def get_phylogeny ( binary: str, infile : str, infile_format : str, args : Optional[str] = 'default' ,
```
Comment (nachorequeno) [inline mevolib/inference/__init__.py]:
```suggestion
def get_phylogeny(binary: str, infile: str, infile_format: str, args: Optional[str] = 'default',
```

File: mevolib/inference/__init__.py [hunk 2]
```diff
@@ -51,19 +68,19 @@
 
 #-------------------------------------------------------------------------------
 
-def get_tools ( ) :
+def get_tools ( ) -> dict :
```
Comment (nachorequeno) [inline mevolib/inference/__init__.py]:
```suggestion
def get_tools() -> dict:
```

File: mevolib/inference/__init__.py [hunk 3]
```diff
@@ -51,19 +68,19 @@
...
     """
     Returns :
         dict
             Dictionary of phylogenetic inference and bootstrapping software
             tools included in the current version of MEvoLib.
     """
-    return ( dict([('inference', list(viewkeys(_PHYLO_TOOL_TO_LIB))),
-                   ('bootstrap', list(viewkeys(_BOOTS_TOOL_TO_LIB)))]) )
+    return ( dict([('inference', list(_PHYLO_TOOL_TO_LIB.keys())),
```
Comment (nachorequeno) [inline mevolib/inference/__init__.py]:
```suggestion
    return ( dict([('inference', list(_PHYLO_TOOL_TO_LIB.keys())),
```

File: mevolib/inference/__init__.py [hunk 4]
```diff
@@ -51,19 +68,19 @@
...
     """
-    return ( dict([('inference', list(viewkeys(_PHYLO_TOOL_TO_LIB))),
-                   ('bootstrap', list(viewkeys(_BOOTS_TOOL_TO_LIB)))]) )
+    return ( dict([('inference', list(_PHYLO_TOOL_TO_LIB.keys())),      
+                   ('bootstrap', list(_BOOTS_TOOL_TO_LIB.keys()))]))
 
 
 
-def get_keywords ( tool ) :
+def get_keywords (tool: str ) -> dict :
```
Comment (nachorequeno) [inline mevolib/inference/__init__.py]:
```suggestion
def get_keywords (tool: str) -> dict:
```

File: mevolib/inference/__init__.py [hunk 5]
```diff
@@ -89,15 +106,15 @@ def get_keywords ( tool ) :
...
+    return ( keyword_dict ) 
 
 
 
-def get_phylogeny ( binary, infile, infile_format, args = 'default',
-                    outfile = None, outfile_format = 'newick',
-                    bootstraps = 0 ) :
+def get_phylogeny ( binary: str, infile : str, infile_format : str, args : Optional[str] = 'default' ,
+                    outfile : Optional[str] = None , outfile_format : Optional[str] = 'newick',
+                    bootstraps : Optional[int] = 0 ) -> (Bio.Phylo.BaseTree,float) :
```
Comment (nachorequeno) [inline mevolib/inference/__init__.py]:
```suggestion
                    bootstraps: Optional[int] = 0) -> (Bio.Phylo.BaseTree, float):
```

File: mevolib/inference/__init__.py [hunk 6]
```diff
@@ -89,15 +106,15 @@ def get_keywords ( tool ) :
...
-    return ( keyword_dict )
+    return ( keyword_dict ) 
 
 
 
-def get_phylogeny ( binary, infile, infile_format, args = 'default',
-                    outfile = None, outfile_format = 'newick',
-                    bootstraps = 0 ) :
+def get_phylogeny ( binary: str, infile : str, infile_format : str, args : Optional[str] = 'default' ,
+                    outfile : Optional[str] = None , outfile_format : Optional[str] = 'newick',
```
Comment (nachorequeno) [inline mevolib/inference/__init__.py]:
```suggestion
                    outfile: Optional[str] = None, outfile_format: Optional[str] = 'newick',
```

File: mevolib/inference/__init__.py [hunk 7]
```diff
@@ -89,15 +106,15 @@ def get_keywords ( tool ) :
         tool_lib_dict = _PHYLO_TOOL_TO_LIB
     else : # tool in _BOOTS_TOOL_TO_LIB
         tool_lib_dict = _BOOTS_TOOL_TO_LIB
-    for key, value in iter(viewitems(tool_lib_dict[tool].KEYWORDS)) :
+    for key,value in tool_lib_dict.items() :
         keyword_dict[key] = ' '.join(value)
-    return ( keyword_dict )
+    return ( keyword_dict )
```
Comment (nachorequeno) [inline mevolib/inference/__init__.py]:
```suggestion
    return keyword_dict
```

File: mevolib/inference/__init__.py [hunk 8]
```diff
@@ -79,7 +96,7 @@ def get_keywords ( tool ) :
             If the tool introduced isn't included in MEvoLib.Inference.
     """
     tool = tool.lower()
-    tool_lib_keys = viewkeys(_PHYLO_TOOL_TO_LIB) | viewkeys(_BOOTS_TOOL_TO_LIB)
+    tool_lib_keys = _PHYLO_TOOL_TO_LIB.keys() | _BOOTS_TOOL_TO_LIB.keys()
     if ( tool not in tool_lib_keys ) :
```
Comment (nachorequeno) [inline mevolib/inference/__init__.py]:
```suggestion
    if tool not in tool_lib_keys:
```

File: mevolib/inference/__init__.py [hunk 9]
```diff
@@ -89,15 +106,15 @@ def get_keywords ( tool ) :
         tool_lib_dict = _PHYLO_TOOL_TO_LIB
     else : # tool in _BOOTS_TOOL_TO_LIB
         tool_lib_dict = _BOOTS_TOOL_TO_LIB
-    for key, value in iter(viewitems(tool_lib_dict[tool].KEYWORDS)) :
+    for key,value in tool_lib_dict.items() :
```
Comment (nachorequeno) [inline mevolib/inference/__init__.py]:
```suggestion
    for key, value in tool_lib_dict.items():
```

File: mevolib/inference/__init__.py [hunk 10]
```diff
@@ -144,7 +161,7 @@ def get_phylogeny ( binary, infile, infile_format, args = 'default',
     * The output file format must be supported by Bio.Phylo.
     """
     # Get the variables associated with the given phylogenetic inference tool
-    bin_path, bin_name = os.path.split(binary)
+    bin_name = os.path.split(binary)
     bin_name = bin_name.lower()
     if ( bin_name in _PHYLO_TOOL_TO_LIB ) :
```
Comment (nachorequeno) [inline mevolib/inference/__init__.py]:
```suggestion
    if bin_name in _PHYLO_TOOL_TO_LIB:
```

File: mevolib/inference/__init__.py [hunk 11]
```diff
@@ -169,14 +186,14 @@ def get_phylogeny ( binary, infile, infile_format, args = 'default',
...
+        output = subprocess.check_output(command, stderr=subprocess.DEVNULL,
                                          universal_newlines=True)
     except subprocess.CalledProcessError as e :
         cleanup(command)
         message = 'Running "{}" raised an exception'.format(' '.join(e.cmd))
         raise RuntimeError(message)
     else :
-        phylogeny, score = get_results(command, output)
+        phylogeny, score = tool_lib.get_results(command, output)
         if ( outfile ) :
```
Comment (nachorequeno) [inline mevolib/inference/__init__.py]:
```suggestion
        if outfile:
```

File: mevolib/inference/_FastTree.py [hunk 12]
```diff
@@ -123,7 +134,7 @@ def get_results ( command, output ) :
 
 
 
-def cleanup ( command ) :
+def cleanup ( command: list ):
```
Comment (nachorequeno) [inline mevolib/inference/_FastTree.py]:
```suggestion
def cleanup (command: list):
```

File: mevolib/inference/_FastTree.py [hunk 13]
```diff
@@ -93,7 +104,7 @@ def gen_args ( args, infile_path, bootstraps ) :
 
 
 
-def get_results ( command, output ) :
+def get_results ( command: list, output: str ) -> Bio.Phylo.BaseTree :
```
Comment (nachorequeno) [inline mevolib/inference/_FastTree.py]:
```suggestion
def get_results (command: list, output: str) -> Bio.Phylo.BaseTree:
```

File: mevolib/inference/__init__.py [hunk 14]
```diff
@@ -32,14 +45,18 @@
...
-from Bio import AlignIO, Phylo
+import Bio.Phylo.BaseTree
+
+from typing import Optional
+
+from io import StringIO
+from Bio import AlignIO,Phylo
+
 
 from . import _FastTree
```
Comment (nachorequeno) [inline mevolib/inference/__init__.py]: @JAlvarezJarreta : I think this suggestion should work. Do you have any preferences about this?

```suggestion
from mevolib.inference import _FastTree, _RAxML
```

File: mevolib/inference/__init__.py [hunk 15]
```diff
@@ -32,14 +45,18 @@
 import tempfile
 import subprocess
 
-from Bio import AlignIO, Phylo
+import Bio.Phylo.BaseTree
+
+from typing import Optional
+
+from io import StringIO
+from Bio import AlignIO,Phylo
```
Comment (nachorequeno) [inline mevolib/inference/__init__.py]:
```suggestion
from Bio import AlignIO, Phylo
```

**Review by nachorequeno — CHANGES_REQUESTED** [review]
Please, apply PEP 8 style. It implies to remove empty spaces and redundant parenthesis (e.g., `return argument_list` instead of `return ( argument_list )`). Usually, IDEs count on a ``Reformat code'' button that automatically applies this coding style. I already did most of the changes for mevolib/inference/__init__.py, but the rest of the documents need to be revised.

File: mevolib/inference/_FastTree.py [hunk 16]
```diff
@@ -33,11 +42,13 @@
 
 import os
 import tempfile
+import Bio.Phylo.BaseTree
```
Comment (JAlvarezJarreta) [inline mevolib/inference/_FastTree.py]: There are three sections on the imports (divided by a blank line):
1. standard library
2. 3rd party libraries
3. local imports

This translates into this import being in the wrong section as well as `from io import StringIO`.

File: mevolib/inference/_FastTree.py [hunk 17]
```diff
@@ -8,12 +8,21 @@
 #
 #-------------------------------------------------------------------------------
 # File :  _FastTree.py
-# Last version :  v1.01 ( 16/Jul/2016 )
+# Last version :  v2.0 ( 25/Sep/2023 )
```
Comment (JAlvarezJarreta) [inline mevolib/inference/_FastTree.py]: This is an ancient header that is no longer in line with the standards. Please refer to other updated files to update the whole header accordingly (including the license that needs to be added in each file).

File: mevolib/inference/_FastTree.py [hunk 18]
```diff
@@ -123,7 +134,7 @@ def get_results ( command, output ) :
 
 
 
-def cleanup ( command ) :
+def cleanup (command: list):
```
Comment (JAlvarezJarreta) [inline mevolib/inference/_FastTree.py]:
```suggestion
def cleanup (command: list) -> None:
```

File: mevolib/inference/_RAxML.py [hunk 19]
```diff
@@ -26,11 +35,12 @@
 import random
 import tempfile
 import shutil
+import Bio.Phylo.BaseTree
```
Comment (JAlvarezJarreta) [inline mevolib/inference/_RAxML.py]: Same as in previous file.

File: mevolib/inference/_RAxML.py [hunk 20]
```diff
@@ -92,7 +102,7 @@ def gen_args ( args, infile_path, bootstraps ) :
 
 
 
-def get_results ( command, output ) :
+def get_results ( command: list, output: str ) -> (Bio.Phylo.BaseTree,float) :
```
Comment (JAlvarezJarreta) [inline mevolib/inference/_RAxML.py]:
```suggestion
def get_results ( command: list, output: str ) -> tuple[Bio.Phylo.BaseTree, float] :
```

File: mevolib/inference/__init__.py [hunk 21]
```diff
@@ -8,12 +8,25 @@
 #
 #-------------------------------------------------------------------------------
 # File :  __init__.py
-# Last version :  v1.10 ( 16/Jul/2016 )
+# Last version :  v2.0 ( 25/Sep/2023 )
```
Comment (JAlvarezJarreta) [inline mevolib/inference/__init__.py]: Same as before.

File: mevolib/inference/__init__.py [hunk 22]
```diff
@@ -32,14 +45,18 @@
 import tempfile
 import subprocess
 
+import Bio.Phylo.BaseTree
```
Comment (JAlvarezJarreta) [inline mevolib/inference/__init__.py]: Same as before.

File: mevolib/inference/__init__.py [hunk 23]
```diff
@@ -51,19 +68,19 @@
 
 #-------------------------------------------------------------------------------
 
-def get_tools ( ) :
+def get_tools() -> dict:
```
Comment (JAlvarezJarreta) [inline mevolib/inference/__init__.py:33]:
```suggestion
def get_tools() -> dict[str, list]:
```
Adding a little bit more information makes it easier to detect bugs.

File: mevolib/inference/__init__.py [hunk 24]
```diff
@@ -89,15 +106,15 @@ def get_keywords ( tool ) :
...
+    return ( keyword_dict ) 
 
 
 
-def get_phylogeny ( binary, infile, infile_format, args = 'default',
-                    outfile = None, outfile_format = 'newick',
-                    bootstraps = 0 ) :
+def get_phylogeny(binary: str, infile: str, infile_format: str, args: Optional[str] = 'default',
+                    outfile : Optional[str] = None , outfile_format : Optional[str] = 'newick',
+                    bootstraps: Optional[int] = 0) -> (Bio.Phylo.BaseTree, float):
```
Comment (JAlvarezJarreta) [inline mevolib/inference/__init__.py]:
```suggestion
                    bootstraps: Optional[int] = 0) -> tuple[Bio.Phylo.BaseTree, float]:
```

File: mevolib/inference/__init__.py [hunk 25]
```diff
@@ -144,7 +161,7 @@ def get_phylogeny ( binary, infile, infile_format, args = 'default',
     * The output file format must be supported by Bio.Phylo.
     """
     # Get the variables associated with the given phylogenetic inference tool
-    bin_path, bin_name = os.path.split(binary)
+    bin_name = os.path.split(binary)
```
Comment (JAlvarezJarreta) [inline mevolib/inference/__init__.py]: Where possible, change the usage of `os.path` by `pathlib` library [[link](https://docs.python.org/3/library/pathlib.html)].

File: mevolib/inference/__init__.py [hunk 26]
```diff
@@ -169,15 +186,15 @@ def get_phylogeny ( binary, infile, infile_format, args = 'default',
     command = [binary] + gen_args(args, infile_path, bootstraps)
     # Run the phylogenetic inference process handling any Runtime exception
     try :
-        output = subprocess.check_output(command, stderr=DEVNULL,
+        output = subprocess.check_output(command, stderr=subprocess.DEVNULL,
```
Comment (JAlvarezJarreta) [inline mevolib/inference/__init__.py]: `subprocess.check_output()` is no longer the recommended approach, but `subprocess.run()` (with the corresponding arguments).

**Review by JAlvarezJarreta — CHANGES_REQUESTED** [review]
Whilst the first iteration looks promising, there is still some work left to do. Have a look at what your colleagues did in other modules to understand the missing elements.

To apply PEP8 comfortably, the most recommended approach is to run [black](https://black.readthedocs.io/) over the files you are editing (remember to use `-l 110` to set max line length to 110 chars).

_Note:_ when agreeing with changes suggested by a reviewer, you can stack them together in GitHub to push them in a single commit instead of generating one commit per suggestion.

**Review by SMoragon — COMMENTED** [review]
Required changes applied and some style issues solved (imports organization, headers and so on).

File: mevolib/inference/__init__.py [hunk 27]
```diff
@@ -203,5 +172,3 @@ def get_phylogeny(binary: str, infile: str, infile_format: str, args: Optional[s
         # log-likelihood score
         return ( phylogeny, score )
```
Comment (nachorequeno) [inline mevolib/inference/__init__.py]:
```suggest
      return phylogeny, score
```

File: mevolib/inference/_FastTree.py [hunk 28]
```diff
@@ -1,74 +1,49 @@
...
 
-
 from mevolib._utils import get_abspath
-from io import StringIO
-
-
-#-------------------------------------------------------------------------------
 
-SPRT_INFILE_FORMATS = ['fasta', 'phylip']
+# -------------------------------------------------------------------------------
```
Comment (nachorequeno) [inline mevolib/inference/_FastTree.py]: Remove `-`'s

File: mevolib/inference/__init__.py (hunk 8 above)
Comment (nachorequeno) [reply mevolib/inference/__init__.py]: Be careful: brackets are back.

File: mevolib/inference/__init__.py (hunk 7 above)
Comment (nachorequeno) [reply mevolib/inference/__init__.py]: Be careful: brackets are back.

File: mevolib/inference/__init__.py (hunk 6 above)
Comment (nachorequeno) [reply mevolib/inference/__init__.py]: Be careful with the white spaces,

**Review by nachorequeno — CHANGES_REQUESTED** [review]
Some suggestions that were applied in previous commit are disappeared. For instance, take a look at `mevolib/inference/__init__.py` file.

**Comment by SMoragon** [comment]
More requested changes solved

File: mevolib/inference/__init__.py [hunk 29]
```diff
@@ -67,7 +67,7 @@ def get_keywords (tool: str) -> dict:
     tool = tool.lower()
     tool_lib_keys = _PHYLO_TOOL_TO_LIB.keys() | _BOOTS_TOOL_TO_LIB.keys()
     if ( tool not in tool_lib_keys ) :
-        raise ValueError('The tool "{}" isn\'t included in ' \
+        raise ValueError('The tool "{}" sisn\'t included in ' \
```
Comment (nachorequeno) [inline mevolib/inference/__init__.py]: Be careful with the typos. In addition, you could replace strings in "{}".format by f-strings.

File: mevolib/inference/__init__.py [hunk 30]
```diff
@@ -67,7 +67,7 @@ def get_keywords (tool: str) -> dict:
     tool = tool.lower()
     tool_lib_keys = _PHYLO_TOOL_TO_LIB.keys() | _BOOTS_TOOL_TO_LIB.keys()
     if ( tool not in tool_lib_keys ) :
```
Comment (nachorequeno) [inline mevolib/inference/__init__.py]:
```suggestion
    if tool not in tool_lib_keys:
```

**Review by nachorequeno — COMMENTED** [review]
I still see several redundant parenthesis (lines 135 and 149) in `__init__.py`. Please, revise this issue in all the Python files: usually, IDE's count on plugins that automatically reformat Python code into a PEP 8 style (i.e., white spaces, brackets, etc.).

Additionally, you could replace strings in `"Hello {0}!".format("Sergio")` by `f-"Hello {name}!"`

**Comment by SMoragon** [comment]
Last style changes

File: mevolib/inference/_FastTree.py [hunk 31]
```diff
@@ -1,142 +1,120 @@
...
-            List of arguments (excluding binary file) to call FastTree.
+        list: List of arguments (excluding binary file) to call FastTree.
     """
-    if ( args in KEYWORDS ) :
+    if args in KEYWORDS:
         argument_list = list(KEYWORDS[args])
-    else : # args not in KEYWORDS
-        argument_list = [arg  for arg in args.split(' ')]
+    else:  # args not in KEYWORDS
+        argument_list = [arg for arg in args.split(" ")]
```
Comment (JAlvarezJarreta) [inline mevolib/inference/_FastTree.py]:
```suggestion
        argument_list = args.split(" ")
```

File: mevolib/inference/_FastTree.py [hunk 32]
```diff
@@ -1,142 +1,120 @@
...
-        list
-            List of arguments (excluding binary file) to call FastTree.
+        list: List of arguments (excluding binary file) to call FastTree.
     """
-    if ( args in KEYWORDS ) :
+    if args in KEYWORDS:
         argument_list = list(KEYWORDS[args])
-    else : # args not in KEYWORDS
-        argument_list = [arg  for arg in args.split(' ')]
+    else:  # args not in KEYWORDS
```
Comment (JAlvarezJarreta) [inline mevolib/inference/_FastTree.py]:
```suggestion
    else:
```
Those types of comments are not useful.

File: mevolib/inference/_FastTree.py [hunk 33]
```diff
@@ -1,142 +1,120 @@
...
     Extract resultant phylogeny and its log-likelihood score from 'output' and
     files generated during the execution of 'command'.
 
     Arguments :
-        command  ( list )
-            FastTree's command line executed.
-        output  ( string )
-            Output from 'command' execution.
+        command: FastTree's command line executed.
+        output:  Output from 'command' execution.
```
Comment (JAlvarezJarreta) [inline mevolib/inference/_FastTree.py]:
```suggestion
        output: Output from 'command' execution.
```

File: mevolib/inference/_FastTree.py [hunk 34]
```diff
@@ -1,142 +1,120 @@
...
-    phylogeny = Phylo.read(StringIO(output), 'newick')
+
+    phylogeny = Phylo.read(StringIO(output), "newick")
     # Read the log file to get the log-likelihood score of the final phylogeny
-    index = command.index('-log') + 1
-    logfile_path = get_abspath(command[index])
-    with open(logfile_path, 'r') as logfile :
+    index = command.index("-log") + 1
+    logfile_path = Path(command[index]).absolute()
+    with open(logfile_path, "r") as logfile:
```
Comment (JAlvarezJarreta) [inline mevolib/inference/_FastTree.py]:
```suggestion
    logfile_path = Path(command[index])
    with logfile_path.open("r") as logfile:
```
`Path()` allows you use `.open()`, which is a nice feature.

File: mevolib/inference/_RAxML.py [hunk 35]
```diff
@@ -1,151 +1,149 @@
...
-        tmpdir_path = tempfile.mkdtemp()
-        argument_list = ['-p', str(seed), '-n', str(pid), '-T', str(NUMCORES),
-                         '-w', tmpdir_path] + KEYWORDS[args]
-    else : # args not in KEYWORDS
-        argument_list = [arg  for arg in args.split(' ')]
+        if tmpdir_path is None:
+            tmpdir_path = tempfile.mkdtemp()
+        argument_list = ["-p", str(seed), "-n", str(pid), "-T", str(NUMCORES), "-w", tmpdir_path] + KEYWORDS[
+            args
+        ]
```
Comment (JAlvarezJarreta) [inline mevolib/inference/_RAxML.py:84]: Is this split in lines produced by `black`?

File: mevolib/inference/_RAxML.py [hunk 36]
```diff
@@ -1,151 +1,149 @@
...
+    else:  # args not in KEYWORDS
+        argument_list = [arg for arg in args.split(" ")]
```
Comment (JAlvarezJarreta) [inline mevolib/inference/_RAxML.py]:
```suggestion
    else:
        argument_list = args.split(" ")
```
Same as before.

File: mevolib/inference/_RAxML.py [hunk 37]
```diff
@@ -1,151 +1,149 @@
...
+    index = command.index("-n") + 1
     outfiles_id = command[index]
     # Get the output directory from 'command' or use current working directory
-    if ( '-w' in command ) :
-        index = command.index('-w') + 1
+    if "-w" in command:
+        index = command.index("-w") + 1
         outdir_path = command[index]
-    else : # '-w' not in command
+    else:  # '-w' not in command
```
Comment (JAlvarezJarreta) [inline mevolib/inference/_RAxML.py]:
```suggestion
    else:
```

File: mevolib/inference/__init__.py [hunk 38]
```diff
@@ -1,103 +1,85 @@
...
+        ValueError: If the tool introduced isn't included in MEvoLib.Inference.
     """
     tool = tool.lower()
-    tool_lib_keys = viewkeys(_PHYLO_TOOL_TO_LIB) | viewkeys(_BOOTS_TOOL_TO_LIB)
-    if ( tool not in tool_lib_keys ) :
-        raise ValueError('The tool "{}" isn\'t included in ' \
-                         'MEvoLib.Inference'.format(tool))
+    tool_lib_keys = _PHYLO_TOOL_TO_LIB.keys() | _BOOTS_TOOL_TO_LIB.keys()
+    if tool not in tool_lib_keys:
+        raise ValueError(f'The tool "{tool}" isn\'t included in ' "MEvoLib.Inference")
```
Comment (JAlvarezJarreta) [inline mevolib/inference/__init__.py]:
```suggestion
        raise ValueError(f"The tool '{tool}' isn't included in 'MEvoLib.Inference'")
```

File: mevolib/inference/__init__.py [hunk 39]
```diff
@@ -1,103 +1,85 @@
...
     # else : # tool in tool_lib_keys
     keyword_dict = {}
-    if ( tool in _PHYLO_TOOL_TO_LIB ) :
+    if tool in _PHYLO_TOOL_TO_LIB:
         tool_lib_dict = _PHYLO_TOOL_TO_LIB
-    else : # tool in _BOOTS_TOOL_TO_LIB
+    else:  # tool in _BOOTS_TOOL_TO_LIB
```
Comment (JAlvarezJarreta) [inline mevolib/inference/__init__.py]:
```suggestion
    else:
```

File: mevolib/inference/__init__.py [hunk 40]
```diff
@@ -107,84 +89,81 @@ def get_phylogeny ( binary, infile, infile_format, args = 'default',
...
     Raises :
-        ValueError
-            If the tool introduced isn't included in MEvoLib.
-        IOError
-            If the input path or the input file provided doesn't exist.
-        RuntimeError
-            If the call to the phylogenetic inference tool command raises an
+        ValueError: If the tool introduced isn't included in MEvoLib.
+        IOError: If the input path or the input file provided doesn't exist.
+        RuntimeError:  If the call to the phylogenetic inference tool command raises an
```
Comment (JAlvarezJarreta) [inline mevolib/inference/__init__.py]:
```suggestion
        RuntimeError: If the call to the phylogenetic inference tool command raises an
```

File: mevolib/inference/__init__.py [hunk 41]
```diff
@@ -107,84 +89,81 @@ def get_phylogeny ( binary, infile, infile_format, args = 'default',
...
         tool_lib = _PHYLO_TOOL_TO_LIB[bin_name]
         sprt_infile_formats = tool_lib.SPRT_INFILE_FORMATS
         gen_args = tool_lib.gen_args
         get_results = tool_lib.get_results
         cleanup = tool_lib.cleanup
-    else : # bin_name not in _PHYLO_TOOL_TO_LIB
-        message = 'The phylogenetic inference tool "{}" isn\'t included in ' \
-                  'MEvoLib.Inference'.format(bin_name)
-        raise ValueError(message)
+    else:  # bin_name not in _PHYLO_TOOL_TO_LIB
```
Comment (JAlvarezJarreta) [inline mevolib/inference/__init__.py]:
```suggestion
    else:
```

**Review by JAlvarezJarreta — CHANGES_REQUESTED** [review]
I have only reviewed so far the `mevolib/inference` modules. Will do a separate review for the tests, although these should go in a separate PR.

File: tests/my_test_Inference__init_.py [hunk 42]
```diff
@@ -0,0 +1,206 @@
...
+from pathlib import Path
+import random
+from typing import ContextManager
+import subprocess
+import pytest
+from pytest import raises
+
+from Bio import Phylo
+from Bio.Phylo import BaseTree
+from Bio.Phylo.Consensus import _BitString
```
Comment (JAlvarezJarreta) [inline tests/my_test_Inference__init_.py]:
```suggestion

from Bio import Phylo
from Bio.Phylo import BaseTree
from Bio.Phylo.Consensus import _BitString
import pytest
from pytest import raises
```

File: tests/my_test_Inference__init_.py [hunk 43]
```diff
@@ -0,0 +1,206 @@
...
+from mevolib.inference import _FastTree as Fast, _RAxML as Rax
+from mevolib._utils import NUMCORES
+
+
+class TestInferenceInit:
+    tmp_dir: Path = Path("tests/inference_init_tmp_dir/").absolute()
+    manifest_dir: Path
+
+    # store and return all _BitStrings
+    def _bitstrs(self, tree):
```
Comment (JAlvarezJarreta) [inline tests/my_test_Inference__init_.py]: Missing type hints and docstring.

File: tests/my_test_Inference__init_.py [hunk 44]
```diff
@@ -0,0 +1,206 @@
...
+        term_names = [term.name for term in tree.get_terminals()]
+        term_names.sort()
+        for clade in tree.get_nonterminals():
+            clade_term_names = [term.name for term in clade.get_terminals()]
+            boolvals = [name in clade_term_names for name in term_names]
+            bitstr = _BitString("".join(map(str, map(int, boolvals))))
+            bitstrs.add(bitstr)
+        return bitstrs
+
+    def compare(self, tree1, tree2):
```
Comment (JAlvarezJarreta) [inline tests/my_test_Inference__init_.py]: Missing type hints and docstring.

File: tests/my_test_Inference__init_.py [hunk 45]
```diff
@@ -0,0 +1,206 @@
...
+        term_names1 = [term.name for term in tree1.get_terminals()]
+        term_names2 = [term.name for term in tree2.get_terminals()]
+        # false if terminals are not the same
+        if set(term_names1) != set(term_names2):
+            return False
+        # true if _BitStrings are the same
+        if self._bitstrs(tree1) == self._bitstrs(tree2):
+            return True
+        else:
+            return False
```
Comment (JAlvarezJarreta) [inline tests/my_test_Inference__init_.py]:
```suggestion
        if set(term_names1) == set(term_names2):
            return self._bitstrs(tree1) == self._bitstrs(tree2)
        return False
```

File: tests/my_test_Inference_RAxML.py [hunk 46]
```diff
@@ -0,0 +1,264 @@
+from pathlib import Path
+import subprocess
+import pytest
+import random
+import os
+
+from Bio import Phylo
+from Bio.Phylo.Consensus import _BitString
```
Comment (JAlvarezJarreta) [inline tests/my_test_Inference_RAxML.py]:
```suggestion
import os
from pathlib import Path
import random
import subprocess

from Bio import Phylo
from Bio.Phylo.Consensus import _BitString
import pytest
```

File: mevolib/inference/_RAxML.py (hunk 35 above)
Comment (SMoragon) [reply mevolib/inference/_RAxML.py:84]: Yes

File: tests/my_test_Inference_RAxML.py [hunk 47]
```diff
@@ -0,0 +1,264 @@
...
+            assert Rax.KEYWORDS[key] == keywords[key]
+
+    @pytest.mark.parametrize(
+        "args, infile_path, bootstraps, tmpdir_path, seed, arg_list",
+        [
+            (
+                "JC+CAT",
+                "tests/Fasta/f001.mafft_linsi.aln",
+                1,
+                tmp_dir,
```
Comment (JAlvarezJarreta) [inline tests/my_test_Inference_RAxML.py]: I struggle to understand the need for this attribute and `tmpdir_path`. Why not just use `tmp_path` fixture of `pytest`?

File: tests/my_test_Inference_RAxML.py [hunk 48]
```diff
@@ -0,0 +1,264 @@
...
+class MockStdOut:
+    def __init__(self, tree_infile_path, tree_outfile_path, info_infile_path, info_outfile_path):
+        with open(tree_infile_path, "r") as fin:
+            self.stdout_tree = fin.read()
+        with open(tree_outfile_path, "w") as fin:
+            fin.write(self.stdout_tree)
+        with open(info_infile_path, "r") as fin:
+            self.stdout_info = fin.read()
+        with open(info_outfile_path, "w") as fin:
+            fin.write(self.stdout_info)
```
Comment (JAlvarezJarreta) [inline tests/my_test_Inference_RAxML.py]: I do not think you need to copy the files, you can just return its content directly when needed.

File: tests/my_test_Inference_FastTree.py [hunk 49]
```diff
@@ -0,0 +1,180 @@
+from pathlib import Path
+import random
+import subprocess
+import pytest
+from io import StringIO
+
+from Bio import Phylo
+from Bio.Phylo.Consensus import _BitString
```
Comment (JAlvarezJarreta) [inline tests/my_test_Inference_FastTree.py]:
```suggestion
from io import StringIO
from pathlib import Path
import random
import subprocess

from Bio import Phylo
from Bio.Phylo.Consensus import _BitString
import pytest
```

**Review by JAlvarezJarreta — CHANGES_REQUESTED** [review]
Whilst not a complete review, some suggestions of some changes.

File: tests/my_test_Inference_RAxML.py (hunk 48 above)
Comment (SMoragon) [reply tests/my_test_Inference_RAxML.py]: The subprocess.run() method does it, and the RAxML get_results function needs to read the file ended at n (random number) placed at raxml_tmp_dir, so it's needed.

File: tests/my_test_Inference_RAxML.py (hunk 47 above)
Comment (SMoragon) [reply tests/my_test_Inference_RAxML.py]: You're right, my bad.

File: mevolib/inference/_FastTree.py [hunk 50]
```diff
@@ -91,26 +91,32 @@ def get_results(command: list, output: str) -> tuple[Bio.Phylo.BaseTree, float]:
...
     logfile_path = Path(command[index])
-    score=-1
+    score = -1
     with logfile_path.open("r") as logfile:
         # It is located at the last line that matches "TreeLogLk.*" pattern
         for line in logfile.readlines():
             if "TreeLogLk" in line:
-                score = float(line.split("\t")[2])  
+                score = float(line.split("\t")[2])
     return (phylogeny, score)
```
Comment (nachorequeno) [inline mevolib/inference/_FastTree.py]:
```suggestion
    return phylogeny, score
```

**Review by JAlvarezJarreta — APPROVED** [review]
Some bits and blobs still need attention, but good enough first version.

**Review by nachorequeno — APPROVED** [review]

**Pull request merged**